src/omnipool/types.rs

```diff
@@ -0,0 +1,365 @@
+use crate::omnipool::types::BalanceUpdate::{Decrease, Increase};
+use num_traits::{CheckedAdd, CheckedSub};
+use sp_arithmetic::{FixedPointNumber, FixedU128};
+use sp_std::ops::{Add, Deref};
+
+/// Asset state representation including asset pool reserve.
+#[derive(Clone, Default, Debug)]
+pub struct AssetReserveState<Balance> {
+    /// Quantity of asset in omnipool
+    pub reserve: Balance,
+    /// Quantity of Hub Asset matching this asset
+    pub hub_reserve: Balance,
+    /// Quantity of LP shares for this asset
+    pub shares: Balance,
+    /// Quantity of LP shares for this asset owned by protocol
+    pub protocol_shares: Balance,
+    /// TVL of asset
+    pub tvl: Balance,
+}
+
+impl<Balance> AssetReserveState<Balance>
+where
+    Balance: Into<<FixedU128 as FixedPointNumber>::Inner> + Copy + CheckedAdd + CheckedSub + Default,
+{
+    /// Calculate price for actual state
+    pub(crate) fn price(&self) -> Option<FixedU128> {
+        FixedU128::checked_from_rational(self.hub_reserve.into(), self.reserve.into())
+    }
+
+    /// Update current asset state with given delta changes.
+    pub fn delta_update(self, delta: &AssetStateChange<Balance>) -> Option<Self> {
+        Some(Self {
+            reserve: (delta.delta_reserve + self.reserve)?,
+            hub_reserve: (delta.delta_hub_reserve + self.hub_reserve)?,
+            shares: (delta.delta_shares + self.shares)?,
+            protocol_shares: (delta.delta_protocol_shares + self.protocol_shares)?,
+            tvl: (delta.delta_tvl + self.tvl)?,
+        })
+    }
+}
+
+/// Indicates whether delta amount should be added or subtracted.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum BalanceUpdate<Balance> {
+    Increase(Balance),
+    Decrease(Balance),
+}
+
+impl<Balance: CheckedAdd + CheckedSub + PartialOrd + Copy + Default> BalanceUpdate<Balance> {
+    /// Merge two update together
+    pub fn merge(self, other: Self) -> Option<Self> {
+        self.checked_add(&other)
+    }
+}
+
+/// The addition operator + for BalanceUpdate.
+///
+/// Panics if overflows in debug builds, in non-debug debug it wraps instead. Use `checked_add` for safe operation.
+///
+/// # Example
+///
+/// ```ignore
```

Comment:
what is the reason to ignore this doc test? Or it would not work due to some dependency?

Because I think it would be handy to have an executable doc test, if possible.